web-services/query/src/main/java/datawave/webservice/query/metric/QueryMetricsWriter.java

```diff
@@ -189,158 +206,145 @@ public long getAge() {
+        try {
+            if (!metricQueue.isEmpty()) {
+                currentFailures.addAll(writeMetricsToHandler(queryMetricHandler, metricQueue));
+                log.debug("wrote " + (metricQueue.size() - currentFailures.size()) + " query metric updates to queryMetricHandler");
+                anySuccess.set(currentFailures.size() < metricQueue.size());
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+            currentFailures.addAll(metricQueue);
+        } finally {
+            if (!metricQueue.isEmpty()) {
+                if (!currentFailures.isEmpty()) {
+                    log.error(currentFailures.size() + " metrics failed write to queryMetricHandler, will retry individually");
+                    currentFailures.stream().forEach(queryMetricHolder -> {
+                        this.failedMetrics.add(new FailureRecord(queryMetricHolder, anySuccess.get()));
+                    });
                 }
-            } catch (Throwable t) {
-                log.error(t.getMessage(), t);
-                log.error("Error writing " + metricQueue.size() + " query metric updates.  Creating new queryMetricHandler.");
-                // error during write or flush, create a new handler so that we can re-try next time
-                queryMetricHandler.reload();
-            } finally {
-                try {
-                    queryMetricHandler.flush();
-                } catch (Throwable t) {
-                    log.error(t.getMessage(), t);
-                } finally {
-                    receivingMetrics.set(false);
+                if (!anySuccess.get()) {
+                    // error during write or flush, create a new handler so that we can re-try next time
+                    log.error("error writing " + metricQueue.size() + " query metric updates - creating new queryMetricHandler");
+                    queryMetricHandler.reload();
                 }
             }
         }
     }
     
-    private void retryFailures() {
-        Iterator<FailureRecord> itr = failedMetricList.iterator();
-        boolean anySuccess = false;
+    private boolean writeFailedMetrics() {
+        Iterator<FailureRecord> itr = failedMetrics.iterator();
+        int successful = 0;
         while (itr.hasNext()) {
             FailureRecord f = itr.next();
-            if (f.getFailuresWhenOthersSucceeded() > 2) {
-                log.error("discarding metric: " + f.getMetric());
-                itr.remove();
-            } else {
+            if (queryMetricsWriterConfiguration.getUseRemoteService()) {
                 try {
                     writeMetricsToRemoteService(Collections.singletonList(f.getMetric()));
                     itr.remove();
-                    anySuccess = true;
-                } catch (Exception e1) {
-                    log.error("metric failed write to remote service, will retry - " + e1.getMessage());
+                    successful++;
```

Comment:
The naming for this variable seems slightly confusing.  An exception is a success?

Reply:
No - if writeMetricsToRemoteService(Collections.singletonList(f.getMetric())) returns with no exception, then we remove that item (itr.remove()) and increment successful.  The catch clause has no code since the metric that was beng written during the exception will remain in failedMetrics and we process the items remaining in that list later.